src/segment-spinner/Segment.ts

```diff
@@ -0,0 +1,229 @@
+            const newValue = segment.text.slice(0, -1);
+            state.setSegment(
+              segment.type,
+              newValue.length === 0
+                ? (segment.minValue as number)
+                : parseNumber(newValue),
+            );
+            setEnteredKeys(newValue);
+          }
+          break;
+        }
+        default:
+          e.preventDefault();
+          e.stopPropagation();
+          if ((isNumeric(e.key) || /^[ap]$/.test(e.key)) && !state.isReadOnly) {
+            onInput(e.key);
+          }
+      }
+    };
+
+    const onInput = (key: string) => {
+      const newValue = enteredKeys + key;
+
+      switch (segment.type) {
+        case "dayPeriod":
+          if (key === "a") {
+            state.setSegment("dayPeriod", 0);
+          } else if (key === "p") {
+            state.setSegment("dayPeriod", 12);
+          }
+          next();
+          break;
+        case "day":
+        case "hour":
+        case "minute":
+        case "second":
+        case "month":
+        case "year": {
+          if (!isNumeric(newValue)) {
+            return;
+          }
+
+          const numberValue = parseNumber(newValue);
+          let segmentValue = numberValue;
+          if (
+            segment.type === "hour" &&
+            state.dateFormatter.resolvedOptions().hour12 &&
+            numberValue === 12
+          ) {
+            segmentValue = 0;
+          } else if (numberValue > (segment.maxValue as number)) {
+            segmentValue = parseNumber(key);
+          }
+
+          state.setSegment(segment.type, segmentValue);
+
+          if (Number(numberValue + "0") > (segment.maxValue as number)) {
+            setEnteredKeys("");
+            next();
+          } else {
+            setEnteredKeys(newValue);
+          }
+          break;
+        }
+      }
+    };
+
+    const onFocus = () => {
+      setEnteredKeys("");
+    };
+
+    const id = useId(htmlProps.id);
+
+    switch (segment.type) {
+      // A separator, e.g. punctuation
+      case "literal":
+        return {
+          role: "presentation",
+          "data-placeholder": false,
+          children: segment.text,
+          ...htmlProps,
+        };
+
+      // These segments cannot be directly edited by the user.
+      case "weekday":
+      case "timeZoneName":
+      case "era":
+        return {
+          role: "presentation",
+          "data-placeholder": true,
+          children: segment.text,
+          ...htmlProps,
+        };
+
+      // Editable segment
+      default:
+        return mergeProps(spinButtonProps, {
+          id,
+          "aria-label": segment.type,
+          "aria-labelledby": `${state["aria-labelledby"]} ${id}`,
+          tabIndex: state.isDisabled ? undefined : 0,
+          onKeyDown,
+          onFocus,
+          onMouseDown: (e: MouseEvent) => e.stopPropagation(),
+          children: segment.text,
+          ...htmlProps,
+        });
+    }
```

Comment:
```suggestion
         return mergeProps(
          {
            ...spinButtonProps,
            id,
            "aria-label": segment.type,
            "aria-labelledby": `${state["aria-labelledby"]} ${id}`,
            tabIndex: state.isDisabled ? undefined : 0,
            onKeyDown,
            onFocus,
            onMouseDown: (e: MouseEvent) => e.stopPropagation(),
            children: segment.text,
          },
          htmlProps,
        );
```

Reply:
This change gives better clarity on which props will be over-written

Reply:
This breaks the behaviour

Reply:
Oh right right !! handlers are not merged in this case. We need to use another mergeProps which is unnecessary.
No change required here.

Reply:
Just use the htmlProps handlers with callAllHandlers thats necessary.